src/gainmap.c

```diff
@@ -314,29 +315,43 @@ avifResult avifImageApplyGainMap(const avifImage * baseImage,
 // A histogram of gain map values is used to remove outliers, which helps with
 // gain map accuracy/compression.
 #define NUM_HISTOGRAM_BUCKETS 1000 // Empirical value
-// Arbitrary max value, roughly equivalent to bringing SDR white to max PQ white (log2(10000/203) ~= 5.62)
-#define BUCKET_MAX_VALUE 6.0f
-#define BUCKET_MIN_VALUE -BUCKET_MAX_VALUE
-#define BUCKET_RANGE (BUCKET_MAX_VALUE - BUCKET_MIN_VALUE)
-#define MAX_OUTLIERS_RATIO 0.001f // 0.1%
+#define MAX_OUTLIERS_RATIO 0.001f  // 0.1%
 
 // Returns the index of the histogram bucket for a given value.
-static int avifValueToBucketIdx(float v)
+static int avifValueToBucketIdx(float v, float bucketMin, float bucketMax)
 {
-    v = AVIF_CLAMP(v, BUCKET_MIN_VALUE, BUCKET_MAX_VALUE);
-    return AVIF_MIN((int)avifRoundf((v - BUCKET_MIN_VALUE) / BUCKET_RANGE * NUM_HISTOGRAM_BUCKETS), NUM_HISTOGRAM_BUCKETS - 1);
+    v = AVIF_CLAMP(v, bucketMin, bucketMax);
+    return AVIF_MIN((int)avifRoundf((v - bucketMin) / (bucketMax - bucketMin) * NUM_HISTOGRAM_BUCKETS), NUM_HISTOGRAM_BUCKETS - 1);
 }
 // Returns the lower end of the value range belonging to the given histogram bucket.
-static float avifBucketIdxToValue(int idx)
+static float avifBucketIdxToValue(int idx, float bucketMin, float bucketMax)
 {
-    return idx * BUCKET_RANGE / NUM_HISTOGRAM_BUCKETS + BUCKET_MIN_VALUE;
+    return idx * (bucketMax - bucketMin) / NUM_HISTOGRAM_BUCKETS + bucketMin;
 }
 
-// Finds the approximate min/max values from the given histogram, excluding outliers.
-// Outliers have at least one empty bucket between them and the rest of the distribution.
-// At most 0.1% of values can be removed as outliers.
-static void avifFindMinMaxWithoutOutliers(int histogram[NUM_HISTOGRAM_BUCKETS], float * rangeMin, float * rangeMax)
+// Finds the approximate min/max values from the given gain map values, excluding outliers.
+// Uses a histogram, with outliers defined as having at least one empty bucket between them
+// and the rest of the distribution. Discards at most 0.1% of values.
+static void avifFindMinMaxWithoutOutliers(const float * gainMapF, int numPixels, float * rangeMin, float * rangeMax)
 {
+    float min = gainMapF[0];
+    float max = gainMapF[0];
+    for (int i = 0; i < numPixels; ++i) {
+        min = AVIF_MIN(min, gainMapF[i]);
+        max = AVIF_MAX(max, gainMapF[i]);
+    }
+
+    const float bucketMin = min;
+    // Use a bucket range of constant size 12 so that each bucket has a fixed size.
+    // 12 is an arbitrary empirical value, equivalent to a [-6, +6] range, where
+    // +6 is roughly equivalent to bringing SDR white to max PQ white (log2(10000/203) ~= 5.62).
+    const float bucketMax = min + 12;
```

Comment:
Why not use [0:12] rather than [min:min+12]? It seems weird to fix the relative maximum but not the absolute minimum.

Also [-6:+6] is mentioned but the used range is [min:min+12]. Is min expected to be around -6?

Reply:
It was intentional but the intention was not very clear :) I reworked the code to make it clearer hopefully that the idea is to have buckets of fixed size.